lib/recursive-methods.rb

```diff
@@ -1,49 +1,80 @@
-# Authoring recursive algorithms. Add comments including time and space complexity for each method.
-
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because it makes as many calls as the number is large
+# Space complexity: O(n) because the stack holds as many calls as the number is large
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 1 || n == 0
+  raise ArgumentError if n < 0
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), 'n' being the length of the string
+# Space complexity: O(n) because the stack holds as many calls as the string is long
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <=1
+  return s[-1] + reverse(s[0..-2])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); it's actually 1/2 n because the method is called half the times of the length of the string, but the constant falls away
+# Space complexity: O(n); same as above, 1/2 n that becomes n
 def reverse_inplace(s)
```

Comment:
1.  This is not in place
1.  This has the same issues as the above for time/space complexity.